### DISCUSSION TITLE
Converting an image to a RGB 555 palette for use in Sega Saturn (PS1 era / midnineties console) homebrew.

### DISCUSSION THREAD

**Discussion by Xenos-Interactive** [ImageMagick/ImageMagick#5160] [Help]
Hello fellow ImageMagick users,

I work as an 3D-artist mainly in Sega Saturn homebrew,
Textures often need comply to a RGB555 (not 565, the last bit is used differently) palette.
I cannot explain how the console processes the color internally (I believe it involves a 16bit integer).
However the custom converters built for the Sega Saturn homebrew crush the colors by dropping
the last 3 bits afaik. Meaning for an RGB888 image the second hex value for each channel is either
0 or 8 (1000 or 0000 in binary).

What I would like to do is convert a 24bit image to a palette of 15 bits conforming to this standard.
Afterwards I would reduce these colors again to a palette of either 256 or 16 colors (as many textures
use 16 or 256 colors out a 15bit palette).

Is there any way to do this. I know this is very specific.

Best regards,
Xenos

Linktree:

https://linktr.ee/TeamStarlane

**Comment by snibgo** [comment]
What version of IM, on what platform? For the version, please show the text output from "magick -version".

What format is the desired output? This might be PNG or BMP or some other format that includes a header, or a plain binary file with pixels only and no header. Are the three 5-bit values combined into one 16-bit value? If so, which end has the extra bit, and what should that be set to?

Do the files have a palette? This would mean the pixel values are indexes into an internal table of colours.

It would help if you can link to sample RGB555 files, and/or a document that defines the format.

**Comment by urban-warrior** [comment]
The BMP format supports a RGB555 subtype, use the `-define bmp:subtype=RGB555` define.  We currently do not support a raw RGB555 image format.

**Comment by snibgo** [comment]
To expand slightly on @urban-warrior's post: depending on the answers to my questions, IM probably has no trivial method for reading and writing the desired format. But it can probably be done with slightly complicated commands, by creating a single-channel image from some arithmetic on the three colour channels.

**Comment by ponut64** [comment]
The bitwise format of Saturn 16-bit RGB color is:

Where "N" is the most significant bit:
N B B B B B G G G G G R R R R R
The "N" bit is used differently depending on where it is in the system.

**Comment by snibgo** [comment]
Okay, that answers some of my questions. What about the others?

**Comment by Xenos-Interactive** [comment]
What Ponut says is correct, but that is after conversion from a 24bit targa image for internal use in the compiled iso.
The desired format is arbitrary, i believe, as long as all colors fall into the rgb555 range.
A BMP file is fine, i can just convert it to a 24bit targa file.
If the files need a palette, like 256 or 16 colors, those still need to fall in the RGB555 space.
So I just reduce to that number of colors and then convert back to 24bit.
The conversion tool takes care of limiting it to a palette for internal usage.

Are the three 5-bit values combined into one 16-bit value? If so, which end has the extra bit, and what should that be set to?
> Yes they are, but it's important what the converter expects and what is used in the end.
Basically all colors not conforming to the RGB555 standard will get reduced, which can result in color loss
if the "paletted" (aka paletted and then converted back to RGB888 space) of 16 or 256 colors cannot be properly
reduced to the RGB555 color space.

So taking a 24bit photo (i use photo textures a lot), converting it to RGB555 helps, cause if I reduce that image
to 16 or 256 colors naturally all colors will be preserved during conversion to the internal Saturn format.

And what the converter expects is RGB888 images in steps of 8 hex values so with the latter hex value
of the color being 0 or 8.

**Comment by snibgo** [comment]
> A BMP file is fine, ...

Okay, so see urban-warrior's post:
```
magick infile.ext -define bmp:subtype=RGB555 outfile.bmp
```

**Reply by Xenos-Interactive** [reply]
i'll go test it out

**Reply by Xenos-Interactive** [reply]
btw, is there like a batch conversion option?

**Comment by Xenos-Interactive** [comment]
I just converted to RGB555 X1R5G5B5 and A1R585B5 .bmp with Gimp.
But those colors do not conform to the Saturn Standards, the converter expects:

[bush-round.zip](https://github.com/ImageMagick/ImageMagick/files/8734457/bush-round.zip)
The .jpg is the original photo.

![new-gradient](https://user-images.githubusercontent.com/50543825/169412548-6ac86b9c-8f34-4a51-92c5-d6e3ae040065.png)
![diagrock_palette](https://user-images.githubusercontent.com/50543825/169412571-eefc9167-fcce-4e4b-92ef-a0755cf987cc.png) These are compatible palettes, as i was told to make them.

![image](https://user-images.githubusercontent.com/50543825/169412715-f0af3359-cfe7-4027-be2c-b5695f136e9c.png)
A compliant color. All channels end in 0 or 8.

![image](https://user-images.githubusercontent.com/50543825/169412895-130c8abf-b4b4-4abd-90a7-539f9759c219.png)
A noncompliant color from one of the .bmps.

**Reply by Xenos-Interactive** [reply]
https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=35284
Well according to this, this is not the .bmp subtype you guys are referring to here.

![image](https://user-images.githubusercontent.com/50543825/169413946-a8e42788-4d56-4b2d-8733-022713adbab8.png)

**Comment by Xenos-Interactive** [comment]
I have successfully converted the .jpg above with the command line.
magick infile.ext -define bmp:subtype=RGB555 outfile.bmp
Image Magick is a really obtuse program and makes me want to scream into a pillow...

[outfile.zip](https://github.com/ImageMagick/ImageMagick/files/8734602/outfile.zip)

The colors also don't conform to the Saturn Standard.
![image](https://user-images.githubusercontent.com/50543825/169416750-c8cfc9da-ba72-462e-88bb-807581ac7b76.png)

**Reply by Xenos-Interactive** [reply]
Ok, so what i basically need is: I need a 24bit image with the last 3bits of all channels of every pixel set to 0.
Because as far as my homebrew colleagues know, there is no file format that conforms to Saturn standards.

**Reply by fmw42** [reply]
I am no expert on this, but can you not bit shift and then bit shift back. See RightShift and LeftShift at https://imagemagick.org/script/command-line-options.php#evaluate (Or just do the division and then multiplication)

**Reply by snibgo** [reply]
>I am no expert on this, but can you not bit shift and then bit shift back.

RightShift and LeftShift repeatedly divide or multiply by 2.0, so if HRDI is used, this won't zero any bits.

"-evaluate And" does its work with integers, so does clear bits.

**Comment by snibgo** [comment]
IM can read an input image, process it, and write an output file. You haven't said what the output file should be.

BMP has a "555" format, in which each channel has only 32 values. If that isn't the format you want, please explain what format you do want. As I said above, it would help if you can link to sample files, and/or a document that defines the format.

> btw, is there like a batch conversion option?

Yes, "magick mogrify".

> Ok, so what i basically need is: I need a 24bit image with the last 3bits of all channels of every pixel set to 0.

But what format? BMP, PNG or what?

**Reply by Xenos-Interactive** [reply]
24bit .tga
[IMG_6149_satconv.zip](https://github.com/ImageMagick/ImageMagick/files/8735098/IMG_6149_satconv.zip)
as in this file

i used this .py script to achieve this
[eriks-rgb555-script.zip](https://github.com/ImageMagick/ImageMagick/files/8735105/eriks-rgb555-script.zip)

however the i can't index my image (using gimp or gale) at 16 or 256 colors without changing the colors
meaning it would choose the new 16 or 256 color palette exclusively from the colors present in the image

the format is a 24bit .tga with the last 3 bits of each channel dropped.
what the saturn uses internally after conversion is a different matter.

**Comment by snibgo** [comment]
I assume the IM is Q16. Windows CMD command:
```
magick infile.ext -evaluate And 63488 -evaluate Multiply %[fx:257/256] -depth 8 outfile.tga
```
Note that decimal 63488 = 248*256 is binary 11111000 00000000. The multiply propagates the first byte to the second, to compensate for the reverse action when the output is reduced to 8 bits.

If you want to reduce the image to 256 colours, I suggest doing that *before* reducing the number of bits:
```
magick infile.ext -colors 256 -evaluate And 63488 -evaluate Multiply %[fx:257/256] -depth 8 outfile.tga
```

**Reply by Xenos-Interactive** [reply]
I cannot do that, because i end up with less than 256 colors.
I need the 16 or 256 colors of the palette to be chosen from the original image. :)
It is not supposed to use the average of the color, just erase colors very close to another one.
Until only the set number are left.

What I could do, is quantize from a 15bit .bmp, if it quantizes from that colorspace.
And then cut the bits. Because they should be 8 hex values apart. So I end up with the same number of colors?

**Reply by Xenos-Interactive** [reply]
my approach has worked.
is used gimp to convert the image to 16bit x1r5g5b5 .bmp.
Then I quantized in that colorspace.
Then I used a .py script and cut the last 3 bits of each color.
I ended up with 16 colors.

**Reply by Xenos-Interactive** [reply]
i got news:
cutting the last 3bits from a 15bit .bmp will not result in color loss.
however quantization happens from a 24 bit colorspace not a 15bit one.

so i do end up with color loss.
i tested that by converting a 24bit jpg with 4076 to a 15bit bmp with 850
when i cut the last 3bits, the image retained 850 colors.
however taking the .bmp and quantizing it down to 256, converting back to rgb (still with 256 unique colors present)
and the saving as a 15bit .bmp resulted in 241 colors present.
i thought that to be the result of 24bit colors being present in the reduced image.

a homebrew colleague of mine has written an algorithm that:
"recursively reduces the amount of colors by increasing the maximum distance allowed"
instead of averaging colors to create new ones.

it has been recommended to me, to do that in image magick or some other api
